src/test/conditions/hashlock_test.cpp

```diff
@@ -0,0 +1,309 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2016 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/basics/strHex.h>
+#include <ripple/beast/unit_test.h>
+#include <ripple/conditions/condition.h>
+#include <ripple/conditions/fulfillment.h>
+#include <ripple/conditions/hashlock.h>
+#include <algorithm>
+#include <vector>
+
+namespace ripple {
+namespace cryptoconditions {
+
+class hashlock_test : public beast::unit_test::suite
+{
+    void
+    check (
+        std::vector<std::uint8_t> const& payload,
+        std::string const& fulfillment,
+        std::string const& condition)
+    {
+        auto f = load_fulfillment (fulfillment);
+        auto c = load_condition (condition);
+
+        BEAST_EXPECT (f);
+        BEAST_EXPECT (c);
+
+        if (f && c)
+        {
+            // Ensure that loading works correctly
+            BEAST_EXPECT (to_string (*f) == fulfillment);
+
+            {
+                auto f2 = load_fulfillment (makeSlice(to_blob (*f)));
+                BEAST_EXPECT (f2);
+                BEAST_EXPECT (*f == *f2);
+            }
+
+            BEAST_EXPECT (to_string (*c) == condition);
+
+            {
+                auto c1 = load_condition (makeSlice(to_blob (*c)));
+                BEAST_EXPECT (c1);
+                BEAST_EXPECT (*c == *c1);
+
+                auto c2 = load_condition (to_string (*c));
+                BEAST_EXPECT (c2);
+                BEAST_EXPECT (*c == *c2);
+            }
+
+            // Ensures that the fulfillment generates
+            // the condition correctly:
+            BEAST_EXPECT (f->condition() == c);
+
+            // Check fulfillment
+            BEAST_EXPECT (validate (*f, {}));
+            BEAST_EXPECT (validate (*f, makeSlice(payload)));
+
+            BEAST_EXPECT (validate (*f, *c, {}));
+            BEAST_EXPECT (validate (*f, *c, makeSlice(payload)));
```

Comment:
On line 77, we're using the `payload` parameter as a message. On line 79, we're using the `payload` parameter as the actual payload. Payloads and messages are distinct things, and especially for hashlocks it's easy to get confused. We know the hashlock will validate with any message (including the payload), but it's easy mistake to expect the message to be the preimage, and using the preimage as the message on line 77 reinforces that confusion.

I'd suggest line 77 should use a different non-empty message to check that it validates